src/dune_pkg/workspace.ml

```diff
@@ -70,3 +70,35 @@ module Repository = struct
 
   let opam_url { name = _; url } = url
 end
+
+let dev_tool_path_to_source_dir path =
+  let lock_dir_location =
+    Path.Build.relative Path.Build.root ".dev-tools.locks" |> Path.build
+  in
+  let absolute = Path.to_absolute_filename lock_dir_location in
+  match Path.External.is_descendant ~of_:(Path.External.of_string absolute) path with
+  | false ->
+    Code_error.raise
+      "External path is not pointing to lock dir location"
+      [ "external", Path.External.to_dyn path
+      ; "dev tool lock dir location", Path.to_dyn lock_dir_location
+      ]
+  | true ->
+    let as_string = Path.External.to_string path in
+    (match String.drop_prefix ~prefix:absolute as_string with
+     | None ->
+       (* we checked for descendants before, thus it has to match the prefix *)
+       assert false
+     | Some suffix ->
+       (match String.split ~on:'/' suffix with
+        | "" :: dev_tool_name :: components ->
+          Path.Source.L.relative
+            Path.Source.root
+            ([ "_build"; ".dev-tools.locks"; dev_tool_name ] @ components)
```

Comment:
This kind of matching is done often. Unfortunately it is wrong. We can specify `--build-dir` on the CLI. You should be matching against `Path.build_dir` which gets set appropriately.